glean-core/rlb/src/net/mod.rs

```diff
@@ -0,0 +1,112 @@
+
+mod http_uploader;
+
+/// The number of milliseconds to make the uploader thread, when told to,
+/// by glean-core.
+const THROTTLE_BACKOFF_TIME_MS: u64 = 60_000;
+
+/// A description of a component used to upload pings.
+pub trait PingUploader: std::fmt::Debug + Send + Sync {
+    /// Uploads a ping to a server.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - the URL path to upload the data to.
+    /// * `body` - the serialized text data to send.
+    /// * `headers` - a vector of tuples containing the headers to send with
+    ///   the request, i.e. (Name, Value).
+    fn upload(&self, url: String, body: Vec<u8>, headers: Vec<(String, String)>) -> UploadResult;
+}
+
+/// The logic for uploading pings: this leaves the actual upload mechanism as
+/// a detail of the user-provided object implementing `PingUploader`.
+pub(crate) struct UploadManager {
+    inner: Arc<Inner>,
+}
+
+struct Inner {
+    server_endpoint: String,
+    uploader: Box<dyn PingUploader + 'static>,
+    thread_running: AtomicBool,
+}
+
+impl UploadManager {
+    /// Create a new instance of the upload manager.
+    ///
+    /// # Arguments
+    ///
+    /// * `server_endpoint` -  the server pings are sent to.
+    /// * `new_uploader` - the instance of the uploader used to send pings.
+    pub(crate) fn new(
+        server_endpoint: String,
+        new_uploader: Box<dyn PingUploader + 'static>,
+    ) -> Self {
+        Self {
+            inner: Arc::new(Inner {
+                server_endpoint,
+                uploader: new_uploader,
+                thread_running: AtomicBool::new(false),
+            }),
+        }
+    }
+
+    /// Signals Glean to upload pings at the next best opportunity.
+    pub(crate) fn trigger_upload(&self) {
+        if self.inner.thread_running.load(Ordering::SeqCst) {
+            log::debug!("The upload task is already running.");
+            return;
+        }
+
+        let inner = Arc::clone(&self.inner);
+
+        thread::spawn(move || {
+            // Mark the uploader as running.
+            inner.thread_running.store(true, Ordering::SeqCst);
+
+            loop {
+                let incoming_task = with_glean(|glean| glean.get_upload_task());
+
+                match incoming_task {
+                    PingUploadTask::Upload(request) => {
+                        let doc_id = request.document_id.clone();
+                        let upload_url = format!("{}{}", inner.server_endpoint, request.path);
+                        let headers: Vec<(String, String)> = request.headers.into_iter().collect();
+                        let result = inner.uploader.upload(upload_url, request.body, headers);
+                        // Process the upload response.
+                        with_glean(|glean| glean.process_ping_upload_response(&doc_id, result));
+                    }
+                    PingUploadTask::Wait => {
+                        thread::sleep(Duration::from_millis(THROTTLE_BACKOFF_TIME_MS));
```

Comment:
With the change above this becomes:

```suggestion
                        thread::sleep(THROTTLE_BACKOFF_TIME);
```